### ISSUE TITLE
Maxwell wrongly interprets the database and table on some binlog events

### ISSUE THREAD

**Issue by nmaquet** [zendesk/maxwell#546]
Today, I found out that our ETL pipeline downstream from Maxwell (v1.1.1) contains erroneous data. After investigating, I tracked down the data corruption to maxwell's output topic. Here's the (redacted) erroneous message produced by Maxwell:

```
{\"database\":\"client_B\",\"table\":\"table_Y\",\"type\":\"delete\",\"ts\":1485365041,\"xid\":1208901877,\"commit\":true,\"data\":{\"id\":2769317,\"field_foo\":\"bar_value\",\"utcinserttimestamp\":\"2017-01-25 17:24:01\"}}"
```

I've tracked down the binlog event that triggered the Maxwell message, and neither the database nor the table correspond (obtained by running `mysqlbinlog --verbose mysql-bin.012997 | grep -C 100 1208901877`).

```
# at 431555743
#170125 17:24:01 server id 1  end_log_pos 431555743     Table_map: `client_A`.`table_X` mapped to number 130
#170125 17:24:01 server id 1  end_log_pos 431555810     Delete_rows: table id 130 flags: STMT_END_F

BINLOG '
Md+IWBMBAAAAPQAAAJ8EuRkAAIIAAAAAAAEADHZjX3ZpbGxhZ2VhdQAKY3Jvbl9sb2NrcwADAw8H
AqUABg==
Md+IWBkBAAAAQwAAAOIEuRkAAIIAAAAAAAEAA//4pUEqABxjcm9uX3Byb2Nlc3NfcmVjb21tZW5k
YXRpb25zMd+IWA==
'/*!*/;
### DELETE FROM `client_A`.`table_X`
### WHERE
###   @1=2769317
###   @2='bar_value'
###   @3=1485365041
# at 431555810
#170125 17:24:01 server id 1  end_log_pos 431555837     Xid = 1208901877
COMMIT/*!*/;
```

I've managed to notice the error relatively easily because `bar_value` makes sense for `table_X` on `client_A` but makes absolutely no sense for `table_Y` on `client_B`. 

What could cause Maxwell to get confused about database and tables? How should I proceed in fixing this? Any help would be highly appreciated.

**Comment by osheroff** [comment]
that's very odd.  you get all the fun ones.  To the tables have identical schemas, functionally?  Ie they both contain `(long, varchar, timestamp)`?  Knowing the schema for each table will help think through this one.  

The two possible candidates here are something wrong with table-mapping, or somehow the schema-processing code got screwed up.  Could `table_X` have been created by some mechanism that would cause it to erroneously share state with `table_Y`?  Maybe `CREATE TABLE table_X like table_y`?

**Comment by osheroff** [comment]
the other possible here could be problems with the table-map-events

**Comment by nmaquet** [comment]
Looking at `SHOW CREATE TABLE` for both tables, it looks like the tables have no relationship, apart from the fact that the three first fields have compatible types.

```
CREATE TABLE `table_X` (
  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
  `bar_field_1` varchar(55) DEFAULT NULL,
  `bar_field_2` timestamp NULL DEFAULT CURRENT_TIMESTAMP,
  PRIMARY KEY (`id`),
  KEY `bar_field_1` (`bar_field_1`)
) ENGINE=InnoDB AUTO_INCREMENT=23086 DEFAULT CHARSET=utf8
```
```
CREATE TABLE `table_Y` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `foo_field_01` varchar(255) NOT NULL DEFAULT '',
  `foo_field_02` datetime DEFAULT NULL,
  `foo_field_03` enum('value1','value2','value3','value4','none') DEFAULT 'none',
   ...
  `foo_field_62` datetime DEFAULT NULL,
  `foo_field_63` tinyint(1) DEFAULT '0',
  `foo_field_64` varchar(255) DEFAULT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `noDuplicates` (`foo_field_01`),
) ENGINE=InnoDB AUTO_INCREMENT=44402443 DEFAULT CHARSET=utf8
```

**Comment by nmaquet** [comment]
Also note that we are using `--include_dbs` and `--include_tables` and that neither `client_A` nor `table_X` are in the list of dbs / tables Maxwell should be replicating.

**Comment by osheroff** [comment]
how long has the maxwell instance been running?  do you have binlogs from when it started?

A theory that fits is that mysql has re-used table-ids in the binlogs.  Here https://github.com/zendesk/maxwell/blob/dc52f05017f396219e47c36815817c6d9dc41be0/src/main/java/com/zendesk/maxwell/replication/TableCache.java#L16 it looks like I'm assuming that table-id re-use never happens, but that could well be a faulty assumption on my part, and if maxwell stayed up long enough that the ids wrapped around, or the master flipped, or... something?  It'd be great if you had the data to see if `table_Y` ever was previously assigned table_id 130.

**Comment by nmaquet** [comment]
`mysqlbinlog --verbose mysql-bin.012782 | grep 'mapped to number 130'` yields
```
#170119 19:50:11 server id 1  end_log_pos 18321296      Table_map: `client_B`.`table_Y` mapped to number 130
```

while `mysqlbinlog --verbose mysql-bin.012995 | grep 'mapped to number 130'` yields
```
#170125 16:01:02 server id 1  end_log_pos 18933282      Table_map: `vc_villageau`.`cron_locks` mapped to number 130
```

**Comment by nmaquet** [comment]
I'll try to find out when to switch over happened.

**Comment by osheroff** [comment]
ok, that looks like a smoking gun to me.  i presume that `vc_...` is the unsanitized `client_A`

**Comment by nmaquet** [comment]
Correct, I edited my comment shortly after.

**Comment by nmaquet** [comment]
FYI, our DB is `mysql  Ver 15.1 Distrib 10.0.21-MariaDB, for debian-linux-gnu (x86_64) using readline 5.2`

**Comment by nmaquet** [comment]
The boundary occurs around `mysql-bin.01293{7,8,9}`, see the below:

```
#170123 18:27:34 server id 1  end_log_pos 265305        Table_map: `client_B`.`table_Y` mapped to number 130
root@mariadb:/var/log/mysql# mysqlbinlog --verbose mysql-bin.012937 | grep 'mapped to number 130' | head -n 1
#170123 19:14:30 server id 1  end_log_pos 27561300      Table_map: `client_B`.`table_Y` mapped to number 130
root@mariadb:/var/log/mysql# mysqlbinlog --verbose mysql-bin.012938 | grep 'mapped to number 130' | head -n 1
#170123 19:16:01 server id 1  end_log_pos 64645         Table_map: `client_C`.`table_X` mapped to number 130
root@mariadb:/var/log/mysql# mysqlbinlog --verbose mysql-bin.012939 | grep 'mapped to number 130' | head -n 1
#170123 19:18:01 server id 1  end_log_pos 25609         Table_map: `client_A`.`table_X` mapped to number 130
root@mariadb:/var/log/mysql# mysqlbinlog --verbose mysql-bin.012940 | grep 'mapped to number 130' | head -n 1
#170123 20:44:02 server id 1  end_log_pos 53576         Table_map: `client_A`.`table_X` mapped to number 130
```

`ls -alht` reveals unusual binlog file sizes at the boundary:

```
...
-rw-rw----  1 mysql mysql  513M Jan 23 22:41 mysql-bin.012940
-rw-rw----  1 mysql mysql  513M Jan 23 20:44 mysql-bin.012939
-rw-rw----  1 mysql mysql  142K Jan 23 19:17 mysql-bin.012938
-rw-rw----  1 mysql mysql   28M Jan 23 19:14 mysql-bin.012937
-rw-rw----  1 mysql mysql  513M Jan 23 19:12 mysql-bin.012936
-rw-rw----  1 mysql mysql  514M Jan 23 18:27 mysql-bin.012935
...
```

**Comment by osheroff** [comment]
ok, that's at least better than the fix I was putting in; looks like maybe we just have to flush the table-map-cache on a rotate event.  I wonder if later versions of mysql try harder to map the event consistently; it feels like this should be constantly crashing maxwell.  either that, or table-ids survive restarts, or something.  I guess I need to complete my source-dive and figure it out.

**Comment by osheroff** [comment]
also wondering if maybe you upgraded mariaDB at that boundary?

**Comment by osheroff** [comment]
id reuse feels like it could be something quite version-specific.

**Comment by nmaquet** [comment]
At the boundary, we did not upgrade MariaDB but we restarted it with the following settings changed with higher values:
`innodb_buffer_pool_size`
`innodb_buffer_pool_instances`
`innodb_purge_threads`

**Comment by osheroff** [comment]
ok, I did a source dive.  table_id is a 2**48 value that lives as long as the server lives, but on server restart it can be re-used.  not quite sure how the maxwell process survived the server restart exactly, but I'll put together a patch for the table cache that should fix this.

**Comment by nmaquet** [comment]
Thanks a lot for the insanely fast turnaround time.
I do have a question: what's the appriopriate procedure to restart a Maxwell-replicated DB instance?
Does Maxwell need to be stopped before bouncing the DB?

**Comment by nmaquet** [comment]
@osheroff `python-mysql-replication` handles `ROTATE_EVENT` explicitly by invalidating the table map: https://github.com/noplay/python-mysql-replication/blob/master/pymysqlreplication/binlogstream.py#L412

**Comment by osheroff** [comment]
yeah, processing the rotate event could totally work too; it's a decent proxy for a server bounce, and it seems *very* unlikely that the table_id rotates around 2**48 within the scope of a binlog. 

better from a perf, standpoint too, I'll change the PR to do that. 

re: 

> what's the appropriate procedure to restart a Maxwell-replicated DB instance?  Does Maxwell need to be stopped before bouncing the DB?

yeah, the "ideal" solution is to get maxwell down safely before bouncing the DB -- otherwise maxwell simply has no place to stash the last-produced log position, and you end up duplicating up to about 1 second of data.  Maybe it's a failure scenario that should be better dealt with, though; have maxwell recognize that the server has gone down and have it hold the replicator position, with an optimistic viewpoint that the server is coming back.  I'll put in an issue to explore that option.

**Comment by nmaquet** [comment]
Thanks for the clarification :+1:

**Comment by osheroff** [comment]
fixed in 1.7.2
